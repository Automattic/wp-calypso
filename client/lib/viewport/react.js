/** @format */
/**
 * External dependencies
 */
import React, { forwardRef, useState, useEffect } from 'react';
import { createHigherOrderComponent } from '@wordpress/compose';

/**
 * Internal dependencies
 */
import {
	isWithinBreakpoint,
	subscribeIsWithinBreakpoint,
	MOBILE_BREAKPOINT,
	DESKTOP_BREAKPOINT,
} from '.';

/**
 * React hook for getting the status for a breakpoint and keeping it updated.
 *
 * @param {String} breakpoint The breakpoint to consider.
 *
 * @returns {Boolean} The current status for the breakpoint.
 */
export function useBreakpoint( breakpoint ) {
	const [ state, setState ] = useState( () => ( {
		isActive: isWithinBreakpoint( breakpoint ),
		breakpoint,
	} ) );

	useEffect( () => {
		function handleBreakpointChange( isActive ) {
			setState( prevState => {
				// Ensure we bail out without rendering if nothing changes, by preserving state.
				if ( prevState.isActive === isActive && prevState.breakpoint === breakpoint ) {
					return prevState;
				}
				return { isActive, breakpoint };
			} );
		}

		const unsubscribe = subscribeIsWithinBreakpoint( breakpoint, handleBreakpointChange );
		// The unsubscribe function is the entire cleanup for the effect.
		return unsubscribe;
	}, [ breakpoint ] );

	return breakpoint === state.breakpoint ? state.isActive : isWithinBreakpoint( breakpoint );
}

/**
 * React hook for getting the status for the mobile breakpoint and keeping it
 * updated.
 *
 * @returns {Boolean} The current status for the breakpoint.
 */
export function useMobileBreakpoint() {
	return useBreakpoint( MOBILE_BREAKPOINT );
}

/**
 * React hook for getting the status for the desktop breakpoint and keeping it
 * updated.
 *
 * @returns {Boolean} The current status for the breakpoint.
 */
export function useDesktopBreakpoint() {
	return useBreakpoint( DESKTOP_BREAKPOINT );
}

/**
 * React higher order component for getting the status for a breakpoint and
 * keeping it updated.
 *
 * @param {String} breakpoint The breakpoint to consider.
 *
 * @returns {Function} A function that given a component returns the
 * wrapped component.
 */
export const withBreakpoint = breakpoint =>
	createHigherOrderComponent(
		WrappedComponent =>
			forwardRef( ( props, ref ) => {
				const isActive = useBreakpoint( breakpoint );
				return <WrappedComponent { ...props } isBreakpointActive={ isActive } ref={ ref } />;
			} ),
		'WithBreakpoint'
	);

/**
 * React higher order component for getting the status for the mobile
 * breakpoint and keeping it updated.
 *
 * @param {React.Component|Function} Wrapped The component to wrap.
 *
 * @returns {Function} The wrapped component.
 */
export const withMobileBreakpoint = createHigherOrderComponent(
	WrappedComponent =>
		forwardRef( ( props, ref ) => {
			const isActive = useBreakpoint( MOBILE_BREAKPOINT );
			return <WrappedComponent { ...props } isBreakpointActive={ isActive } ref={ ref } />;
		} ),
	'WithMobileBreakpoint'
);

/**
 * React higher order component for getting the status for the desktop
 * breakpoint and keeping it updated.
 *
 * @param {React.Component|Function} Wrapped The component to wrap.
 *
 * @returns {Function} The wrapped component.
 */
export const withDesktopBreakpoint = createHigherOrderComponent(
	WrappedComponent =>
		forwardRef( ( props, ref ) => {
			const isActive = useBreakpoint( DESKTOP_BREAKPOINT );
			return <WrappedComponent { ...props } isBreakpointActive={ isActive } ref={ ref } />;
		} ),
	'WithDesktopBreakpoint'
);
