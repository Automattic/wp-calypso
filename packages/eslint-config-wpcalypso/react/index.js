/** @format */

module.exports = {
	extends: '../index.js',
	parserOptions: {
		ecmaVersion: 7,
		ecmaFeatures: {
			jsx: true,
		},
		sourceType: 'module',
	},
	plugins: [ 'react', 'react-hooks', 'wpcalypso' ],
	rules: {
		'react/jsx-curly-spacing': [ 2, 'always' ],
		'react/jsx-no-duplicate-props': 2,
		'react/jsx-no-target-blank': 2,
		'react/jsx-no-undef': 2,
		'react/jsx-tag-spacing': 2,
		'react/jsx-uses-react': 2,
		'react/jsx-uses-vars': 2,
		'react/no-danger': 2,
		'react/no-deprecated': 2,
		'react/no-did-mount-set-state': 2,
		'react/no-did-update-set-state': 2,
		'react/no-is-mounted': 2,
		'react/no-string-refs': 2,
		'react/prefer-es6-class': 2,
		'react/react-in-jsx-scope': 2,
		'react-hooks/rules-of-hooks': 2,
		'react-hooks/exhaustive-deps': 1,
	},
};
