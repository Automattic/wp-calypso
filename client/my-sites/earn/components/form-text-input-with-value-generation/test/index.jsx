/**
 * @jest-environment jsdom
 */

import { render, screen, waitFor } from '@testing-library/react';
import FormTextInputWithValueGeneration from '../index';

describe( 'FormTextInputWithValueGeneration', () => {
	test( 'should render', () => {
		const { container } = render( <FormTextInputWithValueGeneration action="Random" /> );
		expect( container ).toMatchSnapshot();
	} );

	test( 'should generate a value when the action button is clicked', async () => {
		let value = '';
		render(
			<FormTextInputWithValueGeneration
				value={ value }
				onAction={ () => {
					value = 'generated value';
				} }
				action="Random"
			/>
		);
		const actionButton = await screen.getByText( 'Random' );
		await actionButton.click();
		await waitFor( async () => {
			expect( value ).toEqual( 'generated value' );
		} );
	} );
} );
