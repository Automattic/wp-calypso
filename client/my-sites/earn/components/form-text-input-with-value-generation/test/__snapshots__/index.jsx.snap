// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormTextInputWithValueGeneration should render 1`] = `
<div>
  <div
    class="form-text-input-with-value-generation"
    role="group"
  >
    <input
      class="form-text-input form-text-input-with-value-generation__input"
      maxlength="10"
      type="text"
      value=""
    />
    <button
      class="components-button form-text-input-with-value-generation__button is-compact"
      type="button"
    >
      Random
    </button>
  </div>
</div>
`;
